dosbox-staging/dosbox-staging
# Integrate MMX emulation

# Description

Integrate DOSBox MMX emulation patch by Kekko, shared on Vogons forum, https://www.vogons.org/viewtopic.php?t=74877. The PR contains additional fixes and merges from DOSBox-X done by @kklobe.

To enable:
- in the config file set `cputype = pentium_mmx`

![example](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/f3a42294-a438-4372-bb5b-8057bdc6d31d)

It's possible that presence of the MMX instruction set will improve performance of some Windows 9x games or multimedia software too - SiSoft Sandra 2001 (https://archive.org/details/san710_zip) under Windows 98, with Pentium processor emulated:

![Pentium](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/9273df24-cd73-407a-9af1-7b6c2f0a906d)

.. and with Pentium MMX emulated:

![Pentium_MMX](https://github.com/dosbox-staging/dosbox-staging/assets/48332137/3059057d-82fd-47d6-a5fa-06ff27fc6f3e)


## Related issues

https://github.com/dosbox-staging/dosbox-staging/issues/461


# Manual testing

- run _Extreme Assault_ game, test with _MMX Technology_ enabled/disabled in the _Graphic Detail_ menu
- run _MMXtest_ tool from https://calsci.com/techSupport.html (requires Windows 9x) - should report MMX tests passed and report a speedup - speedup values are not stable, but they always show performance improvement, at least on my system
- get _Seven Heaven_ demo by _Exceed_ from https://files.scene.org/view/parties/2000/mekkasymposium00/in64/h7-final.zip - should run correctly on emulated Pentium MMX and refuse to run on non-MMX CPUs
- get the _Impulse Tracker_ and try MMX driver, for example launch it with `IT.EXE -Sitsb16.mmx` command, open a module, start playing with F5 function key

# Checklist

I have:
